Update readme (#18)
thejamiegc/OLA2-SofQuality: ReadMe.md

# OLA2: API Testing, Coverage, and Benchmarking
 - Group: **TofuBytes**

>  - Members: **Isak, Jamie & Helena**

Reviewer: Burde vi ikke beholde vores gruppe medlemmer og gruppenavne eller har vi dem et andet sted ?

> 

Reviewer: ```suggestion
 - Group name: **TofuBytes** Members: **Isak, Jamie & Helena**
```

- Focus on:
    - Code readability and maintainability.
    - Adherence to coding standards.
    - Correct use of mocks in tests.
    - Identification of potential bugs.
- Submit written feedback and any suggested code changes.

We always make code review on everyones code. We use github pull requests to review the code, and below are some examples of some reviews we did:

![Code comment 1](Images/codeComment1.png)

![Code comment 2](Images/codeComment2.png)

Especially in this last comment, we can see that no one is 'in charge' of the code, and that we are always working together to make sure what works, what doesnt and so on.
We always check the code for readability and maintainability, and make sure that the code is up to our standards.

### 8. Software Review
- Perform a **Software Review** based on the [Software Review](https://en.wikipedia.org/wiki/Software_review) article:
    - Functional correctness: Does the code do what it is supposed to do?
    - Non-functional aspects (performance, maintainability, etc.).
    - Best practices (code structure, documentation).
- Submit a **300-word report** reflecting on your findings and improvements.

> In relation to functional correctness, the code performs as expected and passes all the test cases. Testing was done using equivalence partitioning and boundary value analysis. 

Reviewer: Vi mangler at få specificeret hvor vi bruger equivalence partitioning & boundy value analysis.

for det er kun i vores validation tests det bliver anvendt / unit tests. det bliver ikke brugt i vores integration_api_tests som reviewers nok ville kommentere på modsiger hvad vi skriver her

> In relation to functional correctness, the code performs as expected and passes all the test cases. Testing was done using equivalence partitioning and boundary value analysis. 

Reviewer: ```suggestion
In relation to functional correctness, the code performs as expected and passes all the test cases. `Validation_test` was added as a way of ensuring our Validation Testing was done using equivalence partitioning and boundary value analysis. 
```

For equivalence partitioning, the inputs were divided into valid and invalid categories, ensuring that the code handled typical cases within these partitions correctly. 
Additionally, boundary value analysis revealed that the code responds appropriately at the edges of input ranges, such as minimum and maximum allowable values. However, further testing could be beneficial to expand the test coverage, especially for complex input patterns or extreme edge cases.

> We made sure that our software was correct, and that it was easy to maintain. We also made sure that the code was structured in a way that was easy to understand, and that the documentation was clear and concise.

Reviewer: er ikke sikker på hvad der bliver ment med dokumentation her er det de få kommentarer inde i test casen? 
```suggestion
We made sure that our software was correct, and that it was easy to maintain. We also made sure that the code was structured in a way that was easy to understand.
```

The code was made to do what it was supposed to do, and if there were anything that were 'too much' or 'too little', we made sure to fix it.

In relation to best practices, we always try to maintain a clean code structure and follow the coding standards. We make sure to name our methods correctly, so everyone knows what the methods actually do.

> We could improve the documentation, and use more comments directly in the code, so that is something we need to work on.

Reviewer: Godt punkt at nævne vi vil improve dokumentation da vi ikke har noget nuværende, også derfor det var forvirrende længere oppe
```suggestion
We strive to improve documentation, comments is a way we want to reflect that directly in the code.
```

### 9. Reflection on Testing and Code Quality
- Write a **300-400 word reflection** on:
    - Impact of static code analysis tools (PMD, JaCoCo).
    - Importance of mocking in unit testing.
    - Value of code and software reviews.
    - Influence of Equivalence Partitioning and Boundary Value Analysis on test design.
- Discuss how these practices helped in identifying and fixing issues early.

Impact of static code analysis tools:

> - We implemented Qodana, which is a static code analysis tool. This tool helped us identify issues, redundant code or unused code, and helped us fix them before they became a problem. The impact of Qodana was good, because we were able to actually see all the issues we had in our code. For us, this helps us understanding what exactly we need to focus on improving in the code, so we for examplem dont have a lot of redundant or unused code.

Reviewer: Lille stavefejl ved 'example' her

> - We implemented Qodana, which is a static code analysis tool. This tool helped us identify issues, redundant code or unused code, and helped us fix them before they became a problem. The impact of Qodana was good, because we were able to actually see all the issues we had in our code. For us, this helps us understanding what exactly we need to focus on improving in the code, so we for examplem dont have a lot of redundant or unused code.

Reviewer: ```suggestion
- We implemented Qodana, which is a static code analysis tool. This tool helped us identify issues, redundant code, unused code amongst other issues. This helped us detect and fix these issues before they became a problem. The impact of Qodana was good, it provided us with good overheard and enabled us to see the issues that required our attention. For us, this helps us gather a greater understanding of what exactly we need to focus on, and to improve upon for future projects.
```

Influence of Equivalence Partitioning and Boundary Value Analysis on test design:
- Equivalence partitioning helped us understanding a bit more, what needed to be tested. With the combination of the boundary value, we implemented more tests than we would usually, and it also helped us focus more on testing than writing the actual code.
    - Equivalence partitioning helped us identify the major categories of input values, and  boundary values helped us ensure that the boundaries of these categories was tested thoroughly.
 
In general, these practices work together to form a robust safety net that helps identify and address issues early in the development process. By using these techniques early on, we could find and resolve problems before they snowball into larger, more complex issues. This not only leads to higher-quality software but also saves time and effort, reducing the need for extensive debugging later in the development cycle.

---

## Deliverables
- Refined Task Service source code.
- Unit tests with mocks.

> - PMD (or FxCop) report.

Reviewer: ```suggestion
- PMD (or FxCop) report.
  - We used the built-in functionality of Qodana in Rider
```

> - JaCoCo (or Coverlet) code coverage report.

Reviewer: ```suggestion
- JaCoCo (or Coverlet) code coverage report.
  - We used the built-in code coverage tool in JetBrains Rider.
```

- Peer review feedback and any code changes.
- Software review reflection.
- Optional: SonarQube report.

---

## Project Summary
- **Task Service API**: A CRUD API for managing tasks with emphasis on testability and code quality.
- **Key Focus**:
    - Mocking for isolating business logic.
    - Static code analysis for early identification of issues.
    - Code and software reviews for ensuring overall quality.
    - Comprehensive testing with Equivalence Partitioning and Boundary Value Analysis.

---

> ## Tools and Technologies

Reviewer: Vi skal lige have det rettet til her så der kun står de ting vi har brugt

C# 

Moq

Hverken coverlet eller det andet der stod(vu brugte bare riders indbyggede code coverage værktøj)

Osv osv

> - **Languages**: Java/C# (or equivalent).

Reviewer: ```suggestion
- **Languages**: C#.
```

> - **Mocking**: Mockito, EasyMock (Java), or Moq (C#).

Reviewer: ```suggestion
- **Mocking**: Moq (C#).
```

> - **Static Code Analysis**: PMD (Java) or FxCop (C#).

Reviewer: ```suggestion
- **Static Code Analysis**: Qodana & Qodana Cloud .
```

> - **Code Coverage**: JaCoCo (Java) or Coverlet (C#).

Reviewer: ```suggestion
- **Code Coverage**: Built-in Code Coverage tool in JetBrains Rider.
```
